src/allmydata/storage_client.py

```diff
@@ -299,7 +299,9 @@ def _make_storage_server(self, server_id, server):
             "pub-{}".format(str(server_id, "ascii")),  # server_id is v0-<key> not pub-v0-key .. for reasons?
         )
 
-        if len(server["ann"].get(ANONYMOUS_STORAGE_NURLS, [])) > 0:
+        if not self.node_config.get_config(
+                "storage", "force_foolscap", default=True, boolean=True,
+        ) and len(server["ann"].get(ANONYMOUS_STORAGE_NURLS, [])) > 0:
```

Comment:
Of course I want to ask for a test... I expect this is a very annoying codepath to test, though.  Should it be an integration test?  But ... how can you even tell if a client used Foolscapor not?  I suppose you could configure a server to either require or forbid Foolscap and then assert that the client either fails or doesn't depending on its configuration... and then write the three tests for the three different cases (launching a new client node and maybe a new server node for each test)...

Sounds like a drag.  Is there a compelling argument for leaving this without an automated test, instead?

Or ... maybe if this were factored outside of `StorageFarmBroker` it could be adequately tested with a simple unit test instead. But I don't know how difficult such a refactoring would be.

Reply:
And, uh, the actual logic looks fine to me - unless we wanted to be able to control Foolscap for client and server independently (recalling that a node may be configured as both a client _and_ a server simultaneously).  Then maybe we could use a different `[client]force_foolscap` or something (but maybe this would break other things (tests?) that assume a single flag controls behavior of both client and server).

Reply:
I manually tested changing the first clause to `False`, and it did start failing some of the HTTP end-to-end tests in `test_istorageserver`, picking Foolscap instead of HTTP. So that's something.

Adding a new flag would require changing existing tests but that seems fine. So maybe I'll do that.